### ISSUE TITLE
No Audio in WebGL Build

### ISSUE THREAD

**Issue by thelucre** [zeh/usfxr#30]
Getting an error and no audio output in a WebGL build. Any ideas?

```
Only custom filters can be played. Please add a custom filter or 
an audioclip to the audiosource (SfxrGameObject-5.28197).
```

**Comment by zeh** [comment]
Hum, I wasn't aware of that - I think a bunch of webgl features are still
experimental. I'll take a look.

**Comment by zeh-workco** [comment]
It seems to be because of the way audio is generated when doing it dynamically - it needs to use a separate thread (for parallel generation) and that seems to be not fully functional [in the current iteration of WebGL support](http://docs.unity3d.com/Manual/webgl-audio.html). I'll take a deeper look at it this weekend.

**Comment by zeh** [comment]
Indeed, `OnAudioFilterRead` is not supported. The error message does hint at the lack of an `AudioClip`. One is generated automatically by usfxr. But even with a real `AudioClip`, the filter is never read.

To test this, I started an empty, usfxr-less project, with this script in a GameObject:

```using UnityEngine;
using System.Collections;

public class GameTest : MonoBehaviour {

	private float val;

	void OnAudioFilterRead(float[] data, int channels) {
		for (var i = 0; i < data.Length; ++i) {
			data[i] = Mathf.Sin(val);
			val += 0.02f;
		}
	}
}```

Then, I (manually) attached an `AudioClip` script with a .wav file playing.

This produces a test tone in any version of the project, as the audio filter simply replaces the samples with a sin curve.

On a WebGL test using Chrome Canary, it just plays the original .wav audio, completely ignoring the filter. It only plays the audio once, by the way, which is a bug in itself.

So, because `OnAudioFilterRead` is not supported, and that's what usfxr relies on to generate audio samples in a separate thread as needed, usfxr cannot properly run.

Now. With that said, the [WebGL Audio support page](http://docs.unity3d.com/Manual/webgl-audio.html) does point at a potential solution: the [`AudioClip.Create` method](http://docs.unity3d.com/ScriptReference/AudioClip.Create.html), which can be used to generate custom audio clips in real time. This would allow us to create and play a sample as before.

The caveat with these are twofold. First, according to the WebGL audio support page, "AudioClip.Create is currently not supported in Chrome due to a bug, which is going to be fixed in a future release". And second, and more importantly, the `stream` parameter of the `AudioClip.Create` method is not allowed. This means that the samples would have to be generated on the main thread... it's the equivalent of pre-caching them, because it can take some precious miliseconds depending on the audio.

So, to summarize:

* It's a limitation of the current WebGL implementation, maybe a limitation of WebGL as whole.
* A potential solution exists, but it has limitations:
  * No parallel generation.
  * It likely won't work for a while in Chrome.

I'll explore that solution though. Considering the current architecture, it should be fairly simple to just use a separate method to generate the audio using `AudioClip.Create` when WebGL is detected.

**Comment by thelucre** [comment]
Awesome, thanks for the detailed summary. Here's a link to the chrome bug:
https://code.google.com/p/chromium/issues/detail?id=361859

And it looks like a related fix might be coming through in the 5.3 release this December. 
http://forum.unity3d.com/threads/audioclip-create.317032/#post-2331917

**Comment by zeh** [comment]
Hey Eric, I've pushed [a commit](https://github.com/zeh/usfxr/commit/0b314f41b5259ea5b216ff98b0121d177d94041e) that is my initial solution to the problem. It just creates the audio all at once using `AudioClip.Create` when playing it for the first time. Also, not sure why the documentation says it's not supported in Chrome, because it's working super well in Chrome (Canary)!

I'll probably work on it a bit more to make sure nothing is leaking, and see if there's a way to make it perform better (a big deal since it has to block the UI to generate the audio). But right now things shouldn't be silent anymore, at least.

**Comment by thelucre** [comment]
Thank you! Confirmed working in Chrome 47, too.

**Comment by nicotroia** [comment]
Thanks @zeh u rock

**Comment by zeh** [comment]
A bit late, but closing this. Glad it's fixed!
